checks.d/mesos_master.py

```diff
@@ -0,0 +1,246 @@
+                status = AgentCheck.CRITICAL
+                msg = "Got %s when hitting %s" % (r.status_code, url)
+            else:
+                status = AgentCheck.OK
+                msg = "Mesos master instance detected at %s " % url
+        except requests.exceptions.Timeout as e:
+            # If there's a timeout
+            self._timeout_event(url, timeout, aggregation_key)
+            msg = "%s seconds timeout when hitting %s" % (timeout, url)
+            status = AgentCheck.CRITICAL
+        except Exception as e:
+            msg = e.message
+            status = AgentCheck.CRITICAL
+        finally:
+            if self.SERVICE_CHECK_NEEDED:
+                self.service_check(self.SERVICE_CHECK_NAME, status, tags=tags,
+                                   message=msg)
+                self.SERVICE_CHECK_NEEDED = False
+            if status is AgentCheck.CRITICAL:
+                self.warning(msg)
+                return None
+
+        return r.json()
+
+    def _get_master_state(self, url, timeout):
+        return self._get_json(url + '/state.json', timeout)
+
+    def _get_master_stats(self, url, timeout):
+        if self.version >= Version('0.22.0'):
+            endpoint = '/metrics/snapshot'
+        else:
+            endpoint = '/stats.json'
+        return self._get_json(url + endpoint, timeout)
+
+    def _get_master_roles(self, url, timeout):
+        return self._get_json(url + '/roles.json', timeout)
+
+    def _check_leadership(self, url, timeout):
+        json = self._get_master_state(url, timeout)
+
+        if json is not None:
+            self.version = Version(json['version'])
+            if json['leader'] == json['pid']:
+                self.leader = True
+        else:
+            self.leader = False
+        return json
+
+    def check(self, instance):
+        if 'url' not in instance:
+            raise Exception('Mesos instance missing "url" value.')
+
+        url = instance['url']
+        instance_tags = instance.get('tags', [])
+        default_timeout = self.init_config.get('default_timeout', 5)
+        timeout = float(instance.get('timeout', default_timeout))
+
+        json = self._check_leadership(url, timeout)
+        if json:
+            tags = ['mesos_cluster:' + json['cluster'], 'mesos_pid:' + json['pid'], 'mesos_id:' + json['id'], 'mesos_node:master'] + instance_tags
+
+            if self.leader:
+                self.GAUGE('mesos.cluster.total_frameworks', len(json['frameworks']), tags=tags)
+
+                for framework in json['frameworks']:
+                    framework_tags = ['framework:' + framework['id'], 'framework_name:' + framework['name']] + tags
+                    self.GAUGE('mesos.framework.total_tasks', len(framework['tasks']), tags=framework_tags)
+                    resources = framework['used_resources']
+                    [v[1](self, v[0], resources[k], tags=framework_tags) for k, v in self.FRAMEWORK_METRICS.iteritems()]
```

Comment:
You shouldn't use list comprehension to call a function. It makes it harder to read compared to something like:

```python
for key_name, (metric_name, metric_func) in self.FRAMEWORK_METRICS.iteritems():
    metric_func(self, metric_name, resources[key_name], tags=framework_tags)
```